[ERROR] Transforming JavaScript decorators to the configured target environment ("chrome87", "edge88", "es2020", "firefox78", "safari14" + 2 overrides) is not supported yet

### Describe the bug


I was trying to create a new project of vite and Lit element with decorators but since version `>=4.4.0` it gives me the following error when I import `@connectate/ct-button`

[ERROR] Transforming JavaScript decorators to the configured target environment ("chrome87", "edge88", "es2020", "firefox78", "safari14" + 2 overrides) is not supported yet

Does the library have an error? Why didn't this happen before?

### Reproduction

https://stackblitz.com/edit/vitejs-vite-yx7ber

### Steps to reproduce

_No response_

### System Info

```shell
System:
    OS: macOS 13.0.1
    CPU: (8) arm64 Apple M1 Pro
    Memory: 65.28 MB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.16.0 - ~/.nvm/versions/node/v18.16.0/bin/node
    npm: 9.5.1 - ~/.nvm/versions/node/v18.16.0/bin/npm
    pnpm: 8.6.1 - ~/.nvm/versions/node/v18.16.0/bin/pnpm
  Browsers:
    Brave Browser: 114.1.52.130
    Chrome: 114.0.5735.248
    Safari: 16.1
  npmPackages:
    vite: ^4.4.7 => 4.4.7
```


### Used Package Manager

pnpm

### Logs

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/src/node/optimizer/index.ts
+++ b/packages/vite/src/node/optimizer/index.ts
@@ -27,7 +27,7 @@ import {
 import { transformWithEsbuild } from '../plugins/esbuild'
 import { ESBUILD_MODULES_TARGET } from '../constants'
 import { esbuildCjsExternalPlugin, esbuildDepPlugin } from './esbuildDepPlugin'
-import { scanImports } from './scan'
+import { resolveTsconfigRaw, scanImports } from './scan'
 import { createOptimizeDepsIncludeResolver, expandGlobIds } from './resolve'
 export {
   initDepsOptimizer,
@@ -713,8 +713,12 @@ async function prepareEsbuildOptimizerRun(
 
   const optimizeDeps = getDepOptimizationConfig(config, ssr)
 
-  const { plugins: pluginsFromConfig = [], ...esbuildOptions } =
-    optimizeDeps?.esbuildOptions ?? {}
+  const {
+    plugins: pluginsFromConfig = [],
+    tsconfig,
+    tsconfigRaw,
+    ...esbuildOptions
+  } = optimizeDeps?.esbuildOptions ?? {}
 
   await Promise.all(
     Object.keys(depsInfo).map(async (id) => {
@@ -806,6 +810,8 @@ async function prepareEsbuildOptimizerRun(
     metafile: true,
     plugins,
     charset: 'utf8',
+    tsconfig,
+    tsconfigRaw: resolveTsconfigRaw(tsconfig, tsconfigRaw),
     ...esbuildOptions,
     supported: {
       'dynamic-import': true,

--- a/packages/vite/src/node/optimizer/scan.ts
+++ b/packages/vite/src/node/optimizer/scan.ts
@@ -3,7 +3,13 @@ import fsp from 'node:fs/promises'
 import path from 'node:path'
 import { performance } from 'node:perf_hooks'
 import glob from 'fast-glob'
-import type { BuildContext, Loader, OnLoadResult, Plugin } from 'esbuild'
+import type {
+  BuildContext,
+  BuildOptions,
+  Loader,
+  OnLoadResult,
+  Plugin,
+} from 'esbuild'
 import esbuild, { formatMessages, transform } from 'esbuild'
 import colors from 'picocolors'
 import type { ResolvedConfig } from '..'
@@ -224,16 +230,7 @@ async function prepareEsbuildScanner(
     logLevel: 'silent',
     plugins: [...plugins, plugin],
     tsconfig,
-    tsconfigRaw:
-      tsconfig || typeof tsconfigRaw === 'string'
-        ? tsconfigRaw
-        : {
-            ...tsconfigRaw,
-            compilerOptions: {
-              experimentalDecorators: true,
-              ...tsconfigRaw?.compilerOptions,
-            },
-          },
+    tsconfigRaw: resolveTsconfigRaw(tsconfig, tsconfigRaw),
     ...esbuildOptions,
   })
 }
@@ -665,4 +662,23 @@ function shouldExternalizeDep(resolvedId: string, rawId: string): boolean {
 
 function isScannable(id: string): boolean {
   return JS_TYPES_RE.test(id) || htmlTypesRE.test(id)
+}
+
+// esbuild v0.18 only transforms decorators when `experimentalDecorators` is set to `true`.
+// To preserve compat with the esbuild breaking change, we set `experimentalDecorators` to
+// `true` by default if it's unset.
+// TODO: Remove this in Vite 5 and check https://github.com/vitejs/vite/pull/13805#issuecomment-1633612320
+export function resolveTsconfigRaw(
+  tsconfig: string | undefined,
+  tsconfigRaw: BuildOptions['tsconfigRaw'],
+): BuildOptions['tsconfigRaw'] {
+  return tsconfig || typeof tsconfigRaw === 'string'
+    ? tsconfigRaw
+    : {
+        ...tsconfigRaw,
+        compilerOptions: {
+          experimentalDecorators: true,
+          ...tsconfigRaw?.compilerOptions,
+        },
+      }
 }

